src/change_stream/event.rs

```diff
@@ -0,0 +1,119 @@
+use crate::coll::Namespace;
+use bson::{Bson, Document};
+use serde::{Deserialize, Serialize};
+
+/// An opaque token used for resuming an interrupted
+/// [`ChangeStream`](../struct.ChangeStream.html).
+///
+/// When starting a new change stream,
+/// [`start_after`](../option/struct.ChangeStreamOptions.html#structfield.start_after)
+/// and [`resume_after`](../option/struct.ChangeStreamOptions.html#structfield.resume_after) fields
+/// on [`ChangeStreamOptions`](../option/struct.ChangeStreamOptions.html) can be specified
+/// with instances of `ResumeToken`.
+///
+/// See the documentation
+/// [here](https://docs.mongodb.com/manual/changeStreams/#change-stream-resume-token) for more
+/// information on resume tokens.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ResumeToken(pub(crate) Bson);
+
+/// A `ChangeStreamEvent` represents a
+/// [change event](https://docs.mongodb.com/manual/reference/change-events/) in the associated change stream.
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[non_exhaustive]
+pub struct ChangeStreamEvent<T> {
+    /// An opaque token for use when resuming an interrupted `ChangeStream`.
+    ///
+    /// See the documentation
+    /// [here](https://docs.mongodb.com/manual/changeStreams/#change-stream-resume-token) for
+    /// more information on resume tokens.
+    ///
+    /// Also see the documentation on [resuming a change
+    /// stream](https://docs.mongodb.com/manual/changeStreams/#resume-a-change-stream).
+    #[serde(rename = "_id")]
+    pub id: ResumeToken,
+
+    /// Describes the type of operation represented in this change notification.
+    pub operation_type: OperationType,
+
+    /// Identifies which collection or database where the event occurred.
+    pub ns: Option<ChangeStreamEventSource>,
+
+    /// The new name for the ns collection.  Only included for `OperationType::Rename`.
+    pub to: Option<Namespace>,
+
+    /// For unsharded collections this contains a single field, id, with the value of the id of the
+    /// document updated.  For sharded collections, this will contain all the components of the
+    /// shard key in order, followed by the id if the id isn’t part of the shard key.
+    pub document_key: Option<Document>,
+
+    /// Contains a description of updated and removed fields in this operation.
+    pub update_description: Option<UpdateDescription>,
+
+    /// For operations of type "insert" and "replace", this key will contain the document being
+    /// inserted, or the new version of the document that is replacing the existing
+    /// document, respectively.
+    ///
+    /// For operations of type "update", when the `ChangeStream's` full document type is
+    /// `UpdateLookup`, this key will contain a copy of the full version of the document from
+    /// some point after the update occurred. If the document was deleted since the updated
+    /// happened, it will be `None`.
+    pub full_document: Option<T>,
```

Comment:
I think this comment could also be improved. Here is the description from the manual:

> The document created or modified by the insert, replace, delete, update operations (i.e. CRUD operations).
>
> For insert and replace operations, this represents the new document created by the operation.
>
> For delete operations, this field is omitted as the document no longer exists.
>
> For update operations, this field only appears if you configured the change stream with fullDocument set to updateLookup. This field then represents the most current majority-committed version of the document modified by the update operation. This document may differ from the changes described in updateDescription if other majority-committed operations modified the document between the original update operation and the full document lookup.

We probably don't need all the details in the last paragraph, but I think including at least the first line as a summary would be nice.

Reply:
Updated.